File: cpp/test/shipment_entity_test.cpp
// Generated basic-flow test for the shipment entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct ShipmentSetup {
  std::shared_ptr<BluefinDecryptxP2peSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static ShipmentSetup shipment_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/shipment/ShipmentTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinDecryptxP2peSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("shipment01"), Value("shipment02"), Value("shipment03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINDECRYPTXP_PE_TEST_SHIPMENT_ENTID", idmap},
    {"BLUEFINDECRYPTXP_PE_TEST_LIVE", Value("FALSE")},
    {"BLUEFINDECRYPTXP_PE_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINDECRYPTXP_PE_TEST_SHIPMENT_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINDECRYPTXP_PE_TEST_LIVE") == Value("TRUE");

  ShipmentSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void shipment_entity_instance() {
  auto testsdk = BluefinDecryptxP2peSDK::testSDK();
  auto ent = testsdk->shipment();
  ASSERT_EQ(ent->getName(), std::string("shipment"), "entity name");
}

static void shipment_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"shipment", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinDecryptxP2peSDK::testSDK(seed, sdkopts);
  auto se = strsdk->shipment();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinDecryptxP2peSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->shipment();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void shipment_entity_basic() {
  auto setup = shipment_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create", "list", "load"}) {
    auto sk = is_control_skipped("entityOp", std::string("shipment.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto shipment_ref01_ent = client->shipment();
  Value shipment_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "shipment"}), "shipment_ref01"));
  if (!shipment_ref01_data.is_map()) shipment_ref01_data = vmap();
  {
    Value shipment_ref01_data_result = shipment_ref01_ent->create(Struct::clone(shipment_ref01_data), Value::undef());
    shipment_ref01_data = Helpers::toMapAny(shipment_ref01_data_result);
    if (!shipment_ref01_data.is_map()) shipment_ref01_data = vmap();
    ASSERT_TRUE(shipment_ref01_data.is_map(), "expected create result to be a map");
    ASSERT_TRUE(!getp(shipment_ref01_data, "id").is_undef(), "expected created entity to have an id");
  }

  // LIST
  Value shipment_ref01_match = vmap();
  Value shipment_ref01_list = shipment_ref01_ent->list(Struct::clone(shipment_ref01_match), Value::undef());
  ASSERT_TRUE(shipment_ref01_list.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(shipment_ref01_list), vmap({{"id", getp(shipment_ref01_data, "id")}}));
    ASSERT_TRUE(!found.empty(), "expected to find created entity in list");
  }

  // LOAD
  Value shipment_ref01_match_dt0 = vmap({{"id", getp(shipment_ref01_data, "id")}});
  Value shipment_ref01_data_dt0_loaded = shipment_ref01_ent->load(Struct::clone(shipment_ref01_match_dt0), Value::undef());
  Value shipment_ref01_data_dt0_load_result = Helpers::toMapAny(shipment_ref01_data_dt0_loaded);
  ASSERT_TRUE(shipment_ref01_data_dt0_load_result.is_map(), "expected load result to be a map");
  ASSERT_EQ_VAL(getp(shipment_ref01_data_dt0_load_result, "id"), getp(shipment_ref01_data, "id"), "expected load result id to match");

}

int main() {
  T_RUN(shipment_entity_instance);
  T_RUN(shipment_entity_stream);
  T_RUN(shipment_entity_basic);
  return sdktest::summary("shipment_entity_test");
}

File: scala/sdktest/ShipmentEntityTest.scala
// Generated basic-flow test for the shipment entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped ShipmentTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefindecryptxp2pesdk.core.{Helpers, BluefinDecryptxP2peSDK}
import voxgig.bluefindecryptxp2pesdk.utility.struct.Struct

object ShipmentEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("shipment.instance") {
      val testsdk = BluefinDecryptxP2peSDK.testSDK()
      val ent = testsdk.shipment(null)
      rep.check("shipment.instance", ent != null, "expected non-null shipment entity")
    }

    rep.scope("shipment.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/shipment/ShipmentTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinDecryptxP2peSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("shipment01", "SHIPMENT01")
      idmap.put("shipment02", "SHIPMENT02")
      idmap.put("shipment03", "SHIPMENT03")
      val now = System.currentTimeMillis()

      // CREATE
      val shipmentRef01Ent = client.shipment(null)
      var shipmentRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.shipment"), "shipment_ref01"))
      val shipmentRef01DataResult = shipmentRef01Ent.create(shipmentRef01Data, null)
      shipmentRef01Data = Helpers.toMapAny(shipmentRef01DataResult match { case e: SdkEntity => e.data(); case o => o })
      rep.check("shipment.create.map", shipmentRef01Data != null, "expected create result to be a map")
      rep.check("shipment.create.id", shipmentRef01Data != null && shipmentRef01Data.get("id") != null, "expected created entity to have an id")

      // LIST
      val shipmentRef01Match = new LinkedHashMap[String, Object]()
      val shipmentRef01ListResult = shipmentRef01Ent.list(shipmentRef01Match, null)
      rep.check("shipment.list.islist", shipmentRef01ListResult.isInstanceOf[JList[?]], "expected list result to be an array, got " + shipmentRef01ListResult)
      val shipmentRef01List = shipmentRef01ListResult.asInstanceOf[JList[Object]]
      val shipmentRef01ListFound = Struct.select(
          SdkTestSupport.entityListToData(shipmentRef01List), SdkTestSupport.om("id" -> shipmentRef01Data.get("id")))
      rep.check("shipment.list.exists", !Struct.isempty(shipmentRef01ListFound), "expected to find created entity in list")

      // LOAD
      val shipmentRef01MatchDt0 = new LinkedHashMap[String, Object]()
      shipmentRef01MatchDt0.put("id", shipmentRef01Data.get("id"))
      val shipmentRef01DataDt0Loaded = shipmentRef01Ent.load(shipmentRef01MatchDt0, null)
      val shipmentRef01DataDt0LoadResult = Helpers.toMapAny(shipmentRef01DataDt0Loaded match { case e: SdkEntity => e.data(); case o => o })
      rep.check("shipment.load.map", shipmentRef01DataDt0LoadResult != null, "expected load result to be a map")
      rep.eq("shipment.load.id", shipmentRef01Data.get("id"), shipmentRef01DataDt0LoadResult.get("id"))
    }
  }
}
